src/exchange/exchange.py

```diff
@@ -158,16 +167,144 @@ def add_tool_use(self, tool_use: ToolUse) -> None:
+
+        # get the start and end indexes of the first checkpoint, use these to remove message
+        first_checkpoint = self.checkpoint_data.checkpoints[0]
+        first_checkpoint_start_index = first_checkpoint.start_index - self.checkpoint_data.message_index_offset
+
+        # check if the first message is part of the first checkpoint
+        if first_checkpoint_start_index == 0:
+            # remove this checkpoint, as it no longer has any messages
+            self.checkpoint_data.pop(0)
+
+        self.messages.pop(0)
+        self.checkpoint_data.message_index_offset += 1
+
+        if len(self.checkpoint_data.checkpoints) == 0:
+            # we've removed all the checkpoints, so we need to reset the message index offset
+            self.checkpoint_data.message_index_offset = 0
+
+    def pop_last_checkpoint(self) -> Tuple[Checkpoint, List[Message]]:
+        """
+        Reverts the exchange back to the last checkpoint, removing associated messages
+        """
+        removed_checkpoint = self.checkpoint_data.checkpoints.pop()
+        # pop messages until we reach the start of the next checkpoint
+        while len(self.messages) > removed_checkpoint.start_index - self.checkpoint_data.message_index_offset:
+            self.messages.pop()
+
+    def pop_first_checkpoint(self) -> Tuple[Checkpoint, List[Message]]:
+        """
+        Pop the first checkpoint from the exchange, removing associated messages
+        """
+        if len(self.checkpoint_data.checkpoints) == 0:
+            raise ValueError("There are no checkpoints to pop")
+        first_checkpoint = self.checkpoint_data.pop(0)
+
+        # remove messages until we reach the start of the next checkpoint
+        messages = []
+        stop_at_index = first_checkpoint.end_index - self.checkpoint_data.message_index_offset
+        for _ in range(stop_at_index + 1):  # +1 because it's inclusive
+            messages.append(self.messages.pop(0))
+            self.checkpoint_data.message_index_offset += 1
+
+        if len(self.checkpoint_data.checkpoints) == 0:
+            # we've removed all the checkpoints, so we need to reset the message index offset
+            self.checkpoint_data.message_index_offset = 0
+        return first_checkpoint, messages
+
+    def prepend_checkpointed_message(self, message: Message, token_count: int) -> None:
+        """Prepend a message to the exchange, updating the checkpoint data"""
+        self.messages.insert(0, message)
+        new_index = max(0, self.checkpoint_data.message_index_offset - 1)
+        self.checkpoint_data.checkpoints.insert(
+            0,
+            Checkpoint(
+                start_index=new_index,
+                end_index=new_index,
+                token_count=token_count,
+            ),
+        )
+        self.checkpoint_data.message_index_offset = new_index
+
+    @property
+    def is_allowed_to_call_llm(self) -> bool:
```

Comment:
This is kinda LLM dependent? You can call gpt4o without any messages for example

Reply:
Hmmm, maybe this belongs in the Provider class 🤔. I'm going to comment out it's usage now and mark it with a `TODO` so we can revisit it at a later point.